### ISSUE TITLE
Client and Client Commander should perhaps implement interfaces to enable mocking for testing

### ISSUE THREAD

**Issue by lbdl** [go-ocf/go-coap#81]
It is very hard to test applications that use the libs as many of the fields/funcions of the Client and other types are structs with unexported fields etc. Nor do they have interfaces so if we nee to test there is no way to mock their behaviour.

I realise this issue is way too broad but I was wondering the reasoning behind some of the unexported fields/non interfaces. I would be happy to do PR's for the changes am making to the library so that I can mock for tests etc if you feel this would be valid.

**Comment by jkralik** [comment]
Hi. Pls could you share one of your test (example) that will be use such interfaces? I want to know where it is limits you.

Sure you can provide PR.

**Comment by lbdl** [comment]
```golang
func TestHandleInclMessage(t *testing.T) {

	validInclPayload := []byte(`{"dId": "00:ff:0a:95:00:ff","dType": 0,"fwVersion":"0.1.45a","ts":"2019-11-18T15:05:05Z","iRel": 22.22,"iAbs": 11.111,"tmpAbs": 33.33,"bLvl": 44.44}`)
	mockWriter := &mocks.TestWriter{}
	devTime, _ := time.Parse(time.RFC3339, "2019-11-18T15:05:05Z")

	type args struct {
		payload []byte
		t       time.Time
	}
	tests := []struct {
		name       string
		args       args
		badPayload bool
	}{
		{
			"Valid JSON payload",
			args{validInclPayload,
				devTime},
			false,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if !tt.badPayload {
				req := coap.Request{
					Msg: coap.NewDgramMessage(coap.MessageParams{
						Type:      coap.NonConfirmable,
						Code:      coap.POST,
						MessageID: 123,
						Token:     nil,
						Payload:   tt.args.payload,
					}),
					Client:   &coap.ClientConn{},
					Ctx:      nil,
					Sequence: 0,
				}

				// we have to monkey patch the internals of the ClientCommander type as there is no way of mocking/exposing them
				// to be fair we dont entirely need it but we do want to log out messages. If we dont stub the RemoteAddr call etc
				// then we get nil pointer panics in the test suites and we do want the code under test
				var c *coap.ClientCommander
				monkey.PatchInstanceMethod(reflect.TypeOf(c), "RemoteAddr", func(cc *coap.ClientCommander) net.Addr {
					return mocks.TestableNetAddress{"udp", "0.0.0.0"}
				})

				mockMessage := new(mocks.TestDeviceMessage)
				mockMessage.On("Validate", string(tt.args.payload)).Return(nil)
				mockMessage.On("ParseJSON", tt.args.payload, tt.args.t).Return(nil)
				//mockMessage.On("Persist").Return(nil)

				mockTime := new(mocks.TestClock)

				//handler := NewINCLHandler()
				handler := DataMessageHandle{}
				handler.Msg = mockMessage
				handler.SysClock = mockTime
				handler.Handle = handler.MessageHandlingFunction()
				handler.Handle(mockWriter, &req)
				mockMessage.AssertExpectations(t)
			} else {

			}

		})
	}
}
```

apologies about the long example, however in this instance for example what I needed was to mock the return of `clientCommander.RemoteAddr()` call.

This is itself is an un exported field of the `ClientConn` struct which I have amended in my fork to make exported.

```golang
type ClientConn struct {
	Srv          *Server
	Client       *Client
	Commander    *ClientCommander
	ShutdownSync chan error
	Multicast    bool
}
```

This code as it stands **nearly** works other than issues with the monkeypatching however, ideally what I would like to be able to do is something like

```golang

type MockClientCommander struct {}

func (m *mockClientCommander)SomeMethodIWantToMock {
return someDataIWantForTestPurposes
}
```
and the same with `Message` etc or somthing similar. I may of course be missing something obvious.

I hope this make some kind of sense ! BTW thanks for the great library! I more than happy to contribute if you would be happy with this but I'm sure you have reasons that I am missing

**Comment by lbdl** [comment]
I've just realised that I am probably approaching this in the wrong way.
Instead of mocking the private functionality of the library I should instead follow the pattern described [here](https://rakyll.org/interface-pollution/) where I create a wrapper interface which I can then mock for my tests

for my case then I need the `RemoteAddr()` to return something for my tests. This function belongs to the `clientCommander` which is held in `ClientConn.commander` as you know. I have been trying to modify the interfaces so as to mock these when all I really want to access is the _functionality_.

So then to get the testable behaviour I desire I have created a wrapping interface

``` golang
type RequestReader interface {
	RemoteAdress() string
}

type reqReader struct {
    req *coap.Request
}

func (r *reqReader) RemoteAddress() string {
	return r.req.Client.RemoteAddr()
}
```

in my test code I can now do

```golang

type mockReqReader struct {
}

func (m *mockReqReader) RemoteAddress() string {
	return "some test string"
}
```

my bad I'm new to Golang. I will close the issue
